README.md
---
title:  Booklet  
author: David Baynard  
date:   10 Jan 2018  
fontfamily:   libertine
abstract: |  
    Reorder PDF pages to form booklet
...

Reorder pages of a pdf for simple booklet printing.

An experiment with [lopdf](https://docs.rs/lopdf/0.14.1/lopdf/) to solve a relatively annoying problem.

Feedback is welcome, on any aspect.

Note: this only corresponds to a single printing strategy.

# What it does

Some printing software can’t reorganise pdfs for booklet form.
This executable (and library) can help.

For example, take an 11 page A4 pdf.
The pages are:

    1 2 3 4 5 6 7 8 9 10 11

When printing as a booklet they should group as follows:

    1 | 2 3 | 4 5 | 6 7 | 8 9 | 10 11 | Blank

On each sheet, then, should be

    6 7 | 8 5
    4 9 | 10 3
    2 11 | Blank 1

and if printed using short edge duplex, this should be the result.

This program therefore rearranges the pages into

    6 7 8 5 4 9 10 3 2 11 Blank 1

generating `Blank` by deleting the contents from page `1`.

# Use

## Install

Clone, then

    > cargo install

Make sure `~/.cargo/bin` is in `$PATH`.

## Run

    > pdfbooklet _infile_ _outfile_

Take care, this will probably clobber _outfile_.

Also, running without arguments will use `stdout` (or `stdin` and `stdout`).

For help,

    > pdfbooklet --help

## Issues

Use the issue tracker.
Submit any files that do not work.

# Future priorities

Priorities are

- [ ] Produce other page patterns for other printers
- [ ] Logging
- [ ] Producing a guide pdf for printer testing
- [ ] 2-up (so the resulting pdf can be printed directly)


CHANGELOG.md
# Changelog

## [0.1.3] — 2022-07-01

-   Switch from structopt to clap — thanks @oylenshpeegul

## [0.1.2] — 2022-06-30

-   Upgrades dependencies — thanks @nikolaiwarner

## [0.1.1] — 2020-06-21

-   Builds without warnings with Cargo 1.4.4
-   Upgrades dependencies

## [0.1.0] — 2018-01-10

-   Initial version

[0.1.3]: https://github.com/dbaynard/booklet/compare/0.1.2...0.1.3
[0.1.2]: https://github.com/dbaynard/booklet/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/dbaynard/booklet/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/dbaynard/booklet/tree/0.1.0


src/lib.rs
//! # Rearranging pdf pages
//!
//! There are two separate function:
//!
//! 1.  calculating the new order for the pages, and
//! 2.  rearranging the pages themselves.

/// For testing, quickcheck is invaluable.
#[cfg(test)]
#[macro_use]
extern crate quickcheck;

/// Read and write pdf files with `lopdf`.
/// The interface is pleasant, but there are some lacking features (at least at time of writing)
/// which are fixed in the `extra` modules.
extern crate lopdf;

/// Numerical processing.
extern crate num;
#[macro_use]
extern crate num_derive;

/// Calculate the new page order.
pub mod calculate;
pub use calculate::*;

/// Actually rearrange the pages.
pub mod rearrange;
pub use rearrange::*;

/// The extra modules supply functionality that is not specific to pdf processing.
mod extra {
    /// Generate error values.
    pub mod error;
}


src/reorder.rs
use num::Unsigned;

#[derive(Debug, FromPrimitive, ToPrimitive, Clone, Copy, PartialEq)]
enum OnLeaf {
    Nil,
    One,
    Two,
    Three,
}

impl OnLeaf {
    fn new(x: u32) -> OnLeaf {
        use num::FromPrimitive;

        FromPrimitive::from_u32((4 - x % 4) % 4)
            .expect("This cannot fail")
    }
}

#[derive(Debug)]
enum Half {
    Former,
    Latter,
}
#[derive(Debug)]
pub struct PageProps {
    leaves: u32,
    pages: u32,
    new_pages: u32,
    start_page: u32,
    blanks: OnLeaf,
}

impl PageProps {
    pub fn new(pgs: &NonZero<u32>) -> PageProps {
        let pages = *pgs.ex();

        // round up for the sheets of paper used
        let leaves = get_leaves(&pgs);

        // four pages per leaf
        let new_pages = leaves * 4;

        // The first page is the middle face, LHS
        let start_page = leaves * 2;

        let blanks = OnLeaf::new(pages);

        PageProps {
            leaves,
            pages,
            new_pages,
            start_page,
            blanks,
        }
    }

    pub fn next_page_no(&self, page: u32) -> u32 {
        let half = if page > self.start_page {
            Half::Latter
        } else {
            Half::Former
        };
        next_page_no(
            self.new_pages,
            half,
            page,
            )
    }

    pub fn print_order<'a>(&'a self) -> Box<Iterator<Item = Option<u32>> + 'a>
    {
        PageList::new(self).print_order()
    }
}

#[derive(Debug)]
pub struct PageList<'a>(Option<u32>, &'a PageProps);

impl<'a> PageList<'a> {
    pub fn new(pp: &'a PageProps) -> PageList<'a> {
        PageList(None, pp)
    }

    // No need for boxes, but can't be implemented yet
    //pub fn print_order<F>(self) -> impl Iterator<Item = Option<u32>>
    pub fn print_order(self) -> Box<Iterator<Item = Option<u32>> + 'a>
    {
        let p = &self.1.pages;
        let f = self.map(move |x| {
            if x > *p {
                None
            } else {
                Some(x)
            }
        });

        Box::new(f)
    }
}

impl<'a> Iterator for PageList<'a> {
    type Item = u32;

    fn next(&mut self) -> Option<u32> {
        match self.0 {
            None => {
                self.0 = Some(self.1.start_page);
            }
            Some(r) => {
                let x = self.1.next_page_no(r);

                if x == 0 {
                    return None;
                };

                self.0 = Some(x);
            },
        };

        self.0
    }
}

#[derive(Debug, PartialEq)]
pub struct NonZero<T>(T);

impl<T: Unsigned> NonZero<T> {
    pub fn new(u: T) -> Option<NonZero<T>> {
        if u.is_zero() {
            None
        } else {
            Some(NonZero(u))
        }
    }

    fn ex(&self) -> &T {
        &self.0
    }
}

fn get_leaves(pages: &NonZero<u32>) -> u32 {
    (pages.ex() - 1) / 4 + 1
}

fn next_page_no(pages: u32, half: Half, page: u32) -> u32 {
    use self::Half::*;

    match (half, page % 2) {
        (Former,0) => pages - page + 1,
        (Latter,0) => pages - page + 1,
        (Former,1) => page - 1,
        (Latter,1) => page + 1,
        (_,_) => panic!("The impossible has happened.")
    }
}

#[cfg(test)]
mod test_get_leaves {
    use super::*;
    use quickcheck::TestResult;

    quickcheck! {
        fn prop_exact_leaves_correct(x: u32) -> TestResult {
            let y = (4 - x % 4) % 4;
            let x = match NonZero::new(x) {
                Some(r) => r,
                None => return TestResult::discard(),
            };
            let z = 4 * get_leaves(&x) - y;
            TestResult::from_bool(*x.ex() == z)
        }
    }

    quickcheck! {
        fn prop_onleaf_new(x: u32) -> TestResult {
            use num::ToPrimitive;

            let once = OnLeaf::new(x);
            let back = match once.to_u32() {
                Some(_r) => x,
                None => return TestResult::discard(),
            };

            let twice = OnLeaf::new(back);

            TestResult::from_bool(once == twice)
        }
    }

    #[test]
    fn test_next_page() {
        let test_ps = PageProps::new(&NonZero(19));

        assert_eq!(11, test_ps.next_page_no(10));
        assert_eq!(12, test_ps.next_page_no(11));
        assert_eq!(09, test_ps.next_page_no(12));
        assert_eq!(08, test_ps.next_page_no(09));
        assert_eq!(13, test_ps.next_page_no(08));

        let pages0: Vec<u32> = vec![10,11,12,9,8,13,14,7,6,15,16,5,4,17,18,3,2,19,1];
        let pages1: Vec<u32> = vec![11,12,9,8,13,14,7,6,15,16,5,4,17,18,3,2,19,20,0];
        let next_pages: Vec<u32> = pages0.iter()
            .map(|x| test_ps.next_page_no(*x)).collect();

        assert_eq!(pages1, next_pages);
    }

    #[test]
    fn test_gen_pages() {
        let test_ps = PageProps::new(&NonZero(19));
        assert_eq!(
            vec![10,11,12,9,8,13,14,7,6,15,16,5,4,17,18,3,2,19,20,1],
            PageList::new(&test_ps).collect::<Vec<u32>>()
        )
    }

    #[test]
    fn test_print_order() {
        let test_ps = PageProps::new(&NonZero(19));
        let test_out = PageList::new(&test_ps).print_order().collect::<Vec<_>>();
        assert_eq!(
            vec![
                Some(10),
                Some(11),
                Some(12),
                Some(9),
                Some(8),
                Some(13),
                Some(14),
                Some(7),
                Some(6),
                Some(15),
                Some(16),
                Some(5),
                Some(4),
                Some(17),
                Some(18),
                Some(3),
                Some(2),
                Some(19),
                None,
                Some(1)],
            test_out
        )
    }
}



src/rearrange.rs
//! Rearrange pages within a pdf

use std::io;
use std::path::Path;
use extra::error::*;

/// `lopdf` handles the pdf processing
use lopdf::Error;
use lopdf::Document;
use lopdf::{Object,ObjectId};
use lopdf::Dictionary;
use std::collections::BTreeMap;

/// The page ordering
use calculate::*;

/// Pages are retrieved as a map
type PagesInfo = BTreeMap<u32, ObjectId>;

/// Rearrange the pages
///
/// TODO rename to rearrange
///
/// - infile: If a path, use the file, otherwise use stdin
/// - outfile: If a path, use the file, otherwise use stdout
pub fn reorder<P>(
    infile: Option<P>,
    outfile: Option<P>
    ) -> lopdf::Result<()>
    where P: AsRef<Path>
{
    // Load the input pdf
    let mut doc = match infile {
        Some(file_name) => Document::load(file_name),
        None => Document::load_from(io::stdin()),
    }?;

    // Calculate page numbers and properties
    let in_pages = doc.get_pages();
    let pp = NonZero::new(in_pages.len() as u32)
        .map(|x| PageProps::new(&x))
        .ok_or(nonzero_error())?;

    // Rearrange the pages
    rewrite_pages(&mut doc, &pp, &in_pages)?;

    // Write the output pdf
    match outfile {
        Some(file_name) => {
            doc.save(file_name).map_err(Error::from)?;
            Ok(())
        },
        None => doc.save_to(&mut io::stdout()).map_err(Error::from),
    }

}

/// Generate new pages, insert and rearrange the resulting pdf pages.
fn rewrite_pages(
    doc: &mut Document,
    pp: &PageProps,
    in_pages: &PagesInfo,
    ) -> lopdf::Result<()>
{
    // Return a vector, rather than an iterator, to allow subsequent mutation of the document.
    //
    // TODO figure out how to return an iterator without comprimising this.
    let new_pages = generate_pages(doc, &pp, &in_pages)
        .map(Object::Reference)
        .collect();

    let pages_dict = pages_location(doc)?;

    let new_count = pp.new_pages() as i64;

    set_pages_dict(pages_dict, new_count, new_pages);

    Ok(())
}

/// Actually mutate the element listing pages of the pdf.
fn set_pages_dict(
    pages_dict: &mut Dictionary,
    new_count: i64,
    new_pages: Vec<Object>,
    )
{
    pages_dict.set(
        "Kids",
         Object::Array(new_pages),
    );

    pages_dict.set(
        "Count",
        Object::Integer(new_count),
    );
}

/// Return a mutable reference to the element of the pdf corresponding to the page ordering.
///
/// TODO write using `Borrow` to be polymorphic in reference mutability?
fn pages_location<'a>(doc: &'a mut Document) -> lopdf::Result<&'a mut Dictionary>
{
    // TODO it should be possible to pass the ‘Pages’ reference to get_object_mut directly
    let pages = doc.catalog()              // Result<&Dictionary>
        .and_then(|cat| cat.get(b"Pages")) // Result<&Object>
        .and_then(Object::as_reference)?;  // Result<&ObjectId>

    doc.get_object_mut(pages)          // Result<&Object>
        .and_then(Object::as_dict_mut) // Result<&mut Dictionary>
}

/// Process the list of pages and add any required blanks
///
/// TODO now that I get what `move` does is it possible to remove it?
fn generate_pages<'a>(
    doc: &'a mut Document,
    pp: &'a PageProps,
    in_pages: &'a PagesInfo,
    ) -> impl Iterator<Item=ObjectId> + 'a
{
    let f = pp.print_order().filter_map(move |original_page| match original_page {
        None => {
            in_pages.get(&1).ok_or(Error::from(invalid("Couldn't get object")))
                .and_then(|&x| doc.get_object(x))
                .and_then(Object::as_dict)
                .map(Dictionary::clone)
                .map(|mut x| {
                    x.remove(b"Contents");
                    x
                })
                .map(|blank| doc.add_object(blank))
                .ok()
        },
        Some(p) => {
            in_pages.get(&p)
                .map(|x| *x)
        },
    });

    Box::new(f)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore]
    /// This test requires the file test.pdf to be present.
    fn test_reorder() {
        let test_out = reorder(
            Some("test.pdf"),
            Some("test-out.pdf"),
        );

        assert_eq!(true, test_out.is_ok())
    }
}


src/extra/lopdf.rs
//! Add functionality missing from `lopdf`.

use lopdf::Document;
use lopdf::{Object,ObjectId};

/// It is not possible to access objects mutably through the provided methods on `Document`. Rather
/// than doing so using the struct fields and `BTreeMap` functions, this trait adds a new method
/// which is similar to `get_object`.
pub trait GetObjectMut {
    fn get_object_mut(&mut self, id: ObjectId) -> Option<&mut Object>;
}

/// This (working) implementation is very similar to that of `get_object` and the changes
/// can be applied upstream. It may be possible to use the `Borrow` trait to write a single
/// function which can handle both mutable and immutable references.
impl GetObjectMut for Document {
    /// Get mutable object by object id, will recursively dereference a referenced object.
    fn get_object_mut(&mut self, id: ObjectId) -> Option<&mut Object> {
        let is_ref;

        if let Some(object) = self.objects.get(&id) {
            is_ref = object.as_reference();
        } else {
            return None
        }

        if let Some(id) = is_ref {
            return self.get_object_mut(id);
        } else {
            return self.objects.get_mut(&id);
        }
    }
}


src/calculate.rs
//! Calculate the order of pages for printing a pdf as a booklet
//!
//! A booklet prints 4 A4 pages in A5, two on each side (2-up).
//!
//! TODO Reverse the calculation (also enables property tests).

/// How many pages are blank on the final leaf?
#[derive(Debug, FromPrimitive, ToPrimitive, Clone, Copy, PartialEq)]
enum OnLeaf {
    Nil,
    One,
    Two,
    Three,
}

impl OnLeaf {
    /// TODO Link to quickcheck test
    ///
    /// Note: the modulo operations are safe for unsigned ints
    fn new(x: u32) -> OnLeaf {
        use num::FromPrimitive;

        FromPrimitive::from_u32((4 - x % 4) % 4)
            .expect("This cannot fail")
    }
}

/// Into which half of the document does a given page fall?
///
/// The half is important for directing the subsequent page.
#[derive(Debug)]
enum Half {
    Former,
    Latter,
}

/// The set of page properties
#[derive(Debug)]
pub struct PageProps {
    /// How many sheets of paper.
    leaves: u32,
    /// How many pages of content.
    pages: u32,
    /// How many pages total, including blanks.
    new_pages: u32,
    /// Which is the first page for the rearranged output.
    start_page: u32,
    /// How many blank pages added at the end.
    ///
    /// TODO This is currently unused.
    blanks: OnLeaf,
}

impl PageProps {
    /// This is only valid for a positive integer so the input is restricted to that.
    pub fn new(pgs: &NonZero<u32>) -> PageProps {
        let pages = *pgs.ex();

        // round up for the sheets of paper used
        let leaves = get_leaves(&pgs);

        // four pages per leaf
        let new_pages = leaves * 4;

        // The first page is the middle face, LHS
        let start_page = leaves * 2;

        let blanks = OnLeaf::new(pages);

        PageProps {
            leaves,
            pages,
            new_pages,
            start_page,
            blanks,
        }
    }

    /// Given an original page number, what original page is next?
    pub fn next_page_no(&self, page: u32) -> u32 {
        let half = if page > self.start_page {
            Half::Latter
        } else {
            Half::Former
        };
        next_page_no(
            self.new_pages,
            half,
            page,
            )
    }

    /// Produce an iterator for the new order of original page numbers
    pub fn print_order<'a>(&'a self) -> impl Iterator<Item = Option<u32>> + 'a
    {
        PageList::new(self).print_order()
    }

    /// Rather than making the `new_pages` field public, provide the same value with a function.
    pub fn new_pages(&self) -> u32 {
        self.new_pages
    }
}

/// TODO Enforce difference between pages and leaves with newtypes
/// TODO Link to quickcheck
fn get_leaves(pages: &NonZero<u32>) -> u32 {
    (pages.ex() - 1) / 4 + 1
}

/// This (private) function implements the logic of the corresponding `next_page_no` method.
///
/// # Errors
///
/// The exhaustivity of the pattern matches cannot be checked by the compiler, hence the panic.
fn next_page_no(pages: u32, half: Half, page: u32) -> u32 {
    use self::Half::*;

    match (half, page % 2) {
        (Former,0) => pages - page + 1,
        (Latter,0) => pages - page + 1,
        (Former,1) => page - 1,
        (Latter,1) => page + 1,
        (_,_) => panic!("The impossible has happened.")
    }
}

/// This type is required to iterate through the page numbers
#[derive(Debug)]
pub struct PageList<'a>(Option<u32>, &'a PageProps);

impl<'a> PageList<'a> {
    /// The start state has no page number, meaning the iterator returns the first page with the
    /// first iteration.
    pub fn new(pp: &'a PageProps) -> PageList<'a> {
        PageList(None, pp)
    }

    /// Any numbers outside the original page number list correspond to blank pages, represented
    /// here as `None`.
    pub fn print_order(self) -> impl Iterator<Item = Option<u32>> + 'a
    {
        let p = &self.1.pages;
        let f = self.map(move |x| {
            if x > *p {
                None
            } else {
                Some(x)
            }
        });

        Box::new(f)
    }
}

impl<'a> Iterator for PageList<'a> {
    type Item = u32;

    /// Set and return the same state
    fn next(&mut self) -> Option<u32> {
        match self.0 {
            None => {
                self.0 = Some(self.1.start_page);
            }
            Some(r) => {
                let x = self.1.next_page_no(r);

                if x == 0 {
                    return None;
                };

                self.0 = Some(x);
            },
        };

        self.0
    }
}

// TODO Move following to own module
use num::Unsigned;

/// Compiler guaranteed positive integers
// TODO Implement Copy?
#[derive(Debug, PartialEq)]
pub struct NonZero<T>(T);

impl<T: Unsigned> NonZero<T> {
    pub fn new(u: T) -> Option<NonZero<T>> {
        if u.is_zero() {
            None
        } else {
            Some(NonZero(u))
        }
    }

    fn ex(&self) -> &T {
        &self.0
    }
}

#[cfg(test)]
mod test_get_leaves {
    use super::*;
    use quickcheck::TestResult;

    quickcheck! {
        fn prop_exact_leaves_correct(x: u32) -> TestResult {
            let y = (4 - x % 4) % 4;
            let x = match NonZero::new(x) {
                Some(r) => r,
                None => return TestResult::discard(),
            };
            let z = get_leaves(&x).wrapping_mul(4).wrapping_sub(y);
            TestResult::from_bool(*x.ex() == z)
        }
    }

    quickcheck! {
        fn prop_onleaf_new(x: u32) -> TestResult {
            use num::ToPrimitive;

            let once = OnLeaf::new(x);
            let back = match once.to_u32() {
                Some(_r) => x,
                None => return TestResult::discard(),
            };

            let twice = OnLeaf::new(back);

            TestResult::from_bool(once == twice)
        }
    }

    #[test]
    fn test_next_page() {
        let test_ps = PageProps::new(&NonZero(19));

        assert_eq!(11, test_ps.next_page_no(10));
        assert_eq!(12, test_ps.next_page_no(11));
        assert_eq!(09, test_ps.next_page_no(12));
        assert_eq!(08, test_ps.next_page_no(09));
        assert_eq!(13, test_ps.next_page_no(08));

        let pages0: Vec<u32> = vec![10,11,12,9,8,13,14,7,6,15,16,5,4,17,18,3,2,19,1];
        let pages1: Vec<u32> = vec![11,12,9,8,13,14,7,6,15,16,5,4,17,18,3,2,19,20,0];
        let next_pages: Vec<u32> = pages0.iter()
            .map(|x| test_ps.next_page_no(*x)).collect();

        assert_eq!(pages1, next_pages);
    }

    #[test]
    fn test_gen_pages() {
        let test_ps = PageProps::new(&NonZero(19));
        assert_eq!(
            vec![10,11,12,9,8,13,14,7,6,15,16,5,4,17,18,3,2,19,20,1],
            PageList::new(&test_ps).collect::<Vec<u32>>()
        )
    }

    #[test]
    fn test_print_order() {
        let test_ps = PageProps::new(&NonZero(19));
        let test_out = PageList::new(&test_ps).print_order().collect::<Vec<_>>();
        assert_eq!(
            vec![
                Some(10),
                Some(11),
                Some(12),
                Some(9),
                Some(8),
                Some(13),
                Some(14),
                Some(7),
                Some(6),
                Some(15),
                Some(16),
                Some(5),
                Some(4),
                Some(17),
                Some(18),
                Some(3),
                Some(2),
                Some(19),
                None,
                Some(1)],
            test_out
        )
    }
}

